Sources/MapboxMaps/Style/StyleSourceManager.swift

```diff
@@ -0,0 +1,184 @@
+        let sourceDictionary = try source.jsonObject(userInfo: [.nonVolatilePropertiesOnly: true])
+        try addSource(withId: id, properties: sourceDictionary)
+
+        // volatile properties have to be set after the source has been added to the style
+        let volatileProperties = try source.jsonObject(userInfo: [.volatilePropertiesOnly: true])
+
+        try setSourceProperties(for: id, properties: volatileProperties)
+    }
+
+    internal func updateGeoJSONSource(withId id: String, geoJSON: GeoJSONObject) throws {
+        guard let sourceInfo = allSourceIdentifiers.first(where: { $0.id == id }),
+              sourceInfo.type == .geoJson else {
+            throw StyleError(message: "Source with id '\(id)' is not found or not a GeoJSONSource.")
+        }
+
+        applyGeoJSONData(data: geoJSON.sourceData, sourceId: id)
+    }
+
+    // MARK: - Untyped API
+
+    internal func addSource(withId id: String, properties: [String: Any]) throws {
+        try handleExpected {
+            return styleManager.addStyleSource(forSourceId: id, properties: properties)
+        }
+    }
+
+    internal func removeSource(withId id: String) throws {
+        try handleExpected {
+            return styleManager.removeStyleSource(forSourceId: id)
+        }
+
+        workItems.removeValue(forKey: id)?.cancel()
+    }
+
+    internal func sourceExists(withId id: String) -> Bool {
+        return styleManager.styleSourceExists(forSourceId: id)
+    }
+
+    internal func sourceProperty(for sourceId: String, property: String) -> StylePropertyValue {
+        return styleManager.getStyleSourceProperty(forSourceId: sourceId, property: property)
+    }
+
+    internal func sourceProperties(for sourceId: String) throws -> [String: Any] {
+        return try handleExpected {
+            return styleManager.getStyleSourceProperties(forSourceId: sourceId)
+        }
+    }
+
+    internal func setSourceProperty(for sourceId: String, property: String, value: Any) throws {
+        try handleExpected {
+            return styleManager.setStyleSourcePropertyForSourceId(sourceId, property: property, value: value)
+        }
+    }
+
+    internal func setSourceProperties(for sourceId: String, properties: [String: Any]) throws {
+        try handleExpected {
+            return styleManager.setStyleSourcePropertiesForSourceId(sourceId, properties: properties)
+        }
+    }
+
+    // MARK: - Async GeoJSON source data parsing
+
+    private func addGeoJSONSource(_ source: GeoJSONSource, id: String) throws {
+        guard let data = source.data else {
+            try addSourceInternal(source, id: id)
+            return
+        }
+
+        if case GeoJSONSourceData.empty = data {
+            try addSourceInternal(source, id: id)
+            return
+        }
+
+        var emptySource = source
+        emptySource.data = .empty
+
+        try addSourceInternal(emptySource, id: id)
```

Comment:
nit: in both 3 cases we are calling to add internal source with either nil or empty data, can we simply it like this?
```
var source = source
let sourceData: GeoJSONSourceData! = source.data

source.data = sourceData.map { _ in GeoJSONSourceData.empty }
try addSourceInternal(source, id: id)

if sourceData != nil {
    applyGeoJSONData(data: sourceData, sourceId: id)
}
```